posthog/api/client_auth.py

```diff
@@ -0,0 +1,126 @@
+from datetime import timedelta
+from typing import Any
+from uuid import uuid4
+
+from django.core.cache import cache
+from django.http import JsonResponse
+from rest_framework.request import Request
+from rest_framework.decorators import action
+from rest_framework import viewsets, permissions, exceptions, serializers
+
+from posthog.api.routing import TeamAndOrgViewSetMixin
+from posthog.jwt import PosthogJwtAudience, encode_jwt
+from posthog.models import User
+
+from rest_framework.throttling import AnonRateThrottle, UserRateThrottle
+
+DEFAULT_CLIENT_AUTHENTICATION_TIME = timedelta(days=7)
+
+
+class ClientAuthenticationUserRateThrottle(UserRateThrottle):
+    scope = "client_authentication_user"
+    rate = "5/minute"
+
+
+class ClientAuthenticationAnonCheckRateThrottle(AnonRateThrottle):
+    scope = "client_authentication_check"
+    rate = "20/minute"  # Clients should check no more than once every 10 seconds
+
+
+class ClientAuthenticationAnonStartRateThrottle(AnonRateThrottle):
+    scope = "client_authentication"
+    rate = "3/minute"
+
+
+class ConfirmAuthSerializer(serializers.Serializer):
+    code: serializers.CharField = serializers.CharField(max_length=128, required=True)
+    verification: serializers.CharField = serializers.CharField(max_length=128, required=True)
+
+
+class ClientAuthenticationViewset(TeamAndOrgViewSetMixin, viewsets.ViewSet):
+    scope_object = "INTERNAL"
+    queryset = User.objects.none()
+    throttle_classes = [ClientAuthenticationUserRateThrottle]
+
+    def list(self, request: Request, *args: Any, **kwargs: Any) -> JsonResponse:
+        code = request.GET.get("code")
+
+        if not code:
+            raise exceptions.ValidationError({"code": "Missing code"})
```

Comment:
can we do a code identifier as well such as:
{"code": 1, "error Message": "Missing code"}
Its easier to match them on the client and retry based on the errors, so the messages can change easily.

Reply:
Wrong endpoint - thats only for the UI. Your endpoint always returns an endpoint of `status: "missing" | "pending" | "authorized"`

Reply:
oh yeah, my bad.

Reply:
btw this work ended up being a great idea as I realised we can use it to make the toolbar auth flow way better 💪